WCOSS2 module and build updates

KateFriedman-NOAA opened the pull request:
In this PR are the following updates for WCOSS2 (being committed to the auth repo feature/ops-wcoss2 branch). This is a courtesy PR for comments ahead of submitting my fork branch changes into the auth repo branch. The full PR will happen when feature/ops-wcoss2 is merged into the operations branch. I will submit this PR if there are no objections or issues left to resolve next Tuesday. Thanks!

- add modulefiles for workflow owned codes (builds tested)
- add WCOSS2 stanzas in module-setup, machine-setup, link, and build scripts (tested)
- update the ufs-weather-model, GLDAS, and UPP to test versions for WCOSS2 (builds tested)

See a comment in related issue (https://github.com/NOAA-EMC/global-workflow/issues/399#issuecomment-921216364) for more details on these incoming changes to feature/ops-wcoss2 branch. No major changes will happen in this feature branch, only minimal updates to port to WCOSS2.

@EdwardSafford-NOAA please see the new WCOSS2 OznMon and RadMon build modulefiles.

The next PR will likely focus on the setup script updates. Rocoto will be installed on WCOSS2 soon so I can test the setup script produced xml when it's ready for use.

KateFriedman-NOAA committed:
Add WCOSS2 block to module-setup.*.inc files

- add if-block for NOAA Cactus and Dogwood to both module-setup.csh.inc
and module-setup.sh.inc
- source modules, purge modules, then load needed envvar/1.0 module to
unlock library stacks

Refs: #399

KateFriedman-NOAA committed:
Add WCOSS2 block to machine-setup.sh

- add if-block for NOAA Cactus and Dogwood
- initialize modules, purge modules, load needed envvar/1.0 module to
unlock library stacks

Refs: #399

KateFriedman-NOAA committed:
Create new WCOSS2 modulefiles for workflow codes

- add WCOSS2 modulefiles for OznMonBuild and RadmonBuild
- add WCOSS2 modulefile for storm_reloc
- add WCOSS2 modulefiles for workflow util codes: enkf_chgres_recenter,
enkf_chgres_recenter_nc, gaussian_sfcanl, gfs_bufr gfs_fbwndgfs,
fv3nc2nemsio, regrid_nemsio, gfs_util, mkgfsawps, rdbfmsua

Refs: #399

KateFriedman-NOAA committed:
Add WCOSS2 if-block to build_gfs_bufrsnd.sh for SIGIO

- the production library installation on WCOSS2 is hpc-stack and
includes updates to some library variable names
- added an if-block to build_gfs_bufrsnd.sh when on WCOSS2 to set the
SIGIO variables to the new hpc-stack versions
- SIGIO_LIB4 in hpc-stack is now just SIGIO_LIB
- SIGIO_INC4 in hpc-stack is now just SIGIO_INC
- need if-block to retain support for pre-hpc-stack in operations branch
on other platforms

Refs: #399

KateFriedman-NOAA committed:
Add WCOSS2 if-block to build_tropcy_NEMS.sh for SIGIO

- the production library installation on WCOSS2 is hpc-stack and
includes updates to some library variable names
- added an if-block to build_tropcy_NEMS.sh when on WCOSS2 to set the
SIGIO variables to the new hpc-stack versions
- SIGIO_LIB4 in hpc-stack is now just SIGIO_LIB
- SIGIO_INC4 in hpc-stack is now just SIGIO_INC
- need if-block to retain support for pre-hpc-stack in operations branch
on other platforms

Refs: #399

KateFriedman-NOAA committed:
Adjust gaussian_sfcanl Makefile CMD copy

- add CMD to end of exec copy to resolve copy error on WCOSS2

Refs: #399

KateFriedman-NOAA committed:
Add WCOSS2 to utility compile scripts

- add WCOSS2 machine option to compile scripts for mkgfsawps,
overgridid, rdbfmsua, and webtitle codes

Refs: #399

KateFriedman-NOAA committed:
Update parallel hdf5 and netcdf module names

- the parallel hdf5 and netcdf modules were renamed to match hpc-stack
naming conventions: hdf5/1.10.6 and netcdf/4.7.4
- update module names in needed modulefiles

Refs: #399

KateFriedman-NOAA committed:
Add module_base for WCOSS2

- create module_base modulefile for WCOSS2
- remaining modules to find are commented out: pm5, util_shared, HPSS,
and cdo

Refs: #399

KateFriedman-NOAA committed:
Slight cleanup of FV3 build commands

- cleanup/compress the commands in build_fv3.sh for building the
ufs-weather-model/FV3 code
- based on recommended commands provided by Dusan Jovic

Refs: #399

KateFriedman-NOAA committed:
Add WCOSS2 to link script

- add "wcoss2" to link script machine checks
- add FIX_DIR path for temporary WCOSS2 fix_nco_gfsv16 set; will update
to official glopara group location when available

Refs: #399

KateFriedman-NOAA committed:
Update checkout script for component test versions

- set checkout script to clone test copies for ufs-weather-model, GLDAS,
and UPP
- tested builds successfully; temporarily updated upp module in UPP
build to be non-official copy from Wen Meng space (upp/8.0.11)

Refs: #399

KateFriedman-NOAA committed:
Removed HPSS module from module_base.wcoss2

- there will be no HPSS module on WCOSS2
- hsi and htar are available in the environment

Refs: #399

EdwardSafford-NOAA reviewed (commented):
The OznMonBuild.wcoss2 and RadMonBuild.wcoss2 files can be deleted from the global_workflow.  All of the monitoring components are within the GSI and enter the workflow through the DA checkout these days.  My apologies for not noticing this earlier.  I'll take a look through the global_workflow to see if there are any more legacy *Mon files that can be removed.

aerorahul commented on modulefiles/module_base.wcoss2:
```diff
@@ -0,0 +1,42 @@
+#%Module######################################################################
+##
+
+module load envvar/1.0
```
this is loaded in `modulefiles/module-setup.sh.inc`
```suggestion
```

KateFriedman-NOAA replied:
~~Removed from both module-setup.sh.inc and module-setup.csh.inc at 0e9541da17b5d451d7fd1c89e68d7b077caa9678. Thanks!~~

KateFriedman-NOAA replied:
done, removed from module_base.wcoss2 and added back into module-setup.*.inc, thanks!

aerorahul commented on modulefiles/module_base.wcoss2:
```diff
@@ -0,0 +1,42 @@
+module load craype/2.7.8
+module load intel/19.1.3.304
+module load cray-mpich/8.1.7
+module load cray-pals
```
curious, what is `cray-pals`?

KateFriedman-NOAA replied:
Per Carolyn's presentation slide on "Running jobs":

```
• PBS Pro uses the Cray Parallel Application Launch Service (PALS)
to launch MPI jobs
• cray-pals is a module that needs to be loaded to access mpiexec
```

I haven't gotten to testing job submission yet, have added the module in anticipation of needing it at runtime.

```
module show cray-pals
----------------------------------------------------------------------------------------------
   /opt/cray/pe/lmod/modulefiles/core/cray-pals/1.0.12.lua:
----------------------------------------------------------------------------------------------
whatis("PALS - Parallel Application Launch Service")
prepend_path("PATH","/opt/cray/pe/pals/1.0.12/bin")
prepend_path("MANPATH","/opt/cray/pe/pals/1.0.12/man")
set_alias("mpirun","mpiexec")
```

aerorahul commented on util/sorc/compile_gfs_util_wcoss.sh:
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
This line is checking for `hera` and the echo statement says `You are on WCOSS: $target`

KateFriedman-NOAA replied:
Removed Hera reference

aerorahul commented on util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh:
```diff
@@ -3,7 +3,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
same here.

KateFriedman-NOAA replied:
Removed Hera reference

KateFriedman-NOAA commented:
> The OznMonBuild.wcoss2 and RadMonBuild.wcoss2 files can be deleted from the global_workflow. All of the monitoring components are within the GSI and enter the workflow through the DA checkout these days. My apologies for not noticing this earlier. I'll take a look through the global_workflow to see if there are any more legacy *Mon files that can be removed.

Thanks @EdwardSafford-NOAA ! I'll open a separate issue to remove those legacy files from the develop branch and will ping you in it for your inputs on that cleanup. 

Regarding removing those legacy files in the operations branch port to WCOSS2....@aerorahul @arunchawla-NOAA are we able to delete files in the transition to WCOSS2 or should we leave things as is if it doesn't need an update for WCOSS2? Thanks!

aerorahul commented on util/sorc/overgridid.fd/compile_overgridid_wcoss.sh:
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
same here.

KateFriedman-NOAA replied:
Removed Hera reference

aerorahul commented on util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh:
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
and here.

KateFriedman-NOAA replied:
Removed Hera reference

aerorahul commented on util/sorc/webtitle.fd/compile_webtitle_wcoss.sh:
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] || [ "$target" = "wcoss2" ] ; then
```
here as well.

KateFriedman-NOAA replied:
Removed Hera reference

aerorahul reviewed (changes requested):
most changes look ok.
Some comments and changes requested.

aerorahul commented:
> > The OznMonBuild.wcoss2 and RadMonBuild.wcoss2 files can be deleted from the global_workflow. All of the monitoring components are within the GSI and enter the workflow through the DA checkout these days. My apologies for not noticing this earlier. I'll take a look through the global_workflow to see if there are any more legacy *Mon files that can be removed.
> 
> Thanks @EdwardSafford-NOAA ! I'll open a separate issue to remove those legacy files from the develop branch and will ping you in it for your inputs on that cleanup.
> 
> Regarding removing those legacy files in the operations branch port to WCOSS2....@aerorahul @arunchawla-NOAA are we able to delete files in the transition to WCOSS2 or should we leave things as is if it doesn't need an update for WCOSS2? Thanks!

If they are not going to be used for building these monitoring utilities, then just don't add them in this PR.
There is no point in adding these files in this PR, not use them, and then remove them in an subsequent PR.

KateFriedman-NOAA committed:
Remove load of envvar/1.0 module in module-setup

- envvar/1.0 module will be loaded by module_base.wcoss2
- removing envvar/1.0 module load from module-setup.*.inc scripts

Refs: #399

KateFriedman-NOAA committed:
Remove envvar/1.0 from module_base, add back to setup

- undo prior commit
- add envvar/1.0 back to module-setup.*.inc
- remove envvar/1.0 from module_base.wcoss2

Refs: #399

KateFriedman-NOAA committed:
Remove unnecessary OznMon/RadMon modulefiles

- the OznMonBuild.wcoss2 and RadMonBuild.wcoss2 are not needed and are
considered legacy since the GSI now handles OznMon/Radmon internally

Refs: #399

KateFriedman-NOAA committed:
Remove Hera reference in util compile scripts

- removed Hera from the check of WCOSS partitions in the util source
code compile scripts

Refs: #399

KateFriedman-NOAA committed:
Add util_shared/1.4.0 module load to module_base.wcoss2

- NCO installed util_shared/1.4.0 module on WCOSS2
- need it in module_base.wcoss2

Refs: #399

KateFriedman-NOAA committed:
Reorganize module load order in wcoss2 modulefiles

- reorganize module load order to meet dependencies and order them based
on groups: compilers, mpi, 3rd party, hdf5, netcdf, nceplibs
- add modules to satisfy prereqs for current modules

Refs: #399

KateFriedman-NOAA committed:
Update gfs_util code build

- remove g2/3.4.1 module from gfs_util.wcoss2; not needed
- reorganize modules in gfs_util.wcoss2 and add nemsio and sigio modules
for w3emc prereqs
- fix rdbfmsua makefile for successful compilation (Boi V provided)

Refs: #399

KateFriedman-NOAA committed:
Update g2 library module to g2/3.4.5 for storm_reloc

- g2/3.4.1 reportedly degrades performance on WCOSS2 (Boi V)
- update g2 module in storm_reloc module file from g2/3.4.1 to g2/3.4.5

Refs: #399

aerorahul commented on modulefiles/module_base.wcoss2:
```diff
@@ -0,0 +1,33 @@
+module load cray-python/3.8.5.1
+module load prod_envir/2.0.4
+module load gempak/7.14.0
+module load nco/4.7.9
```
`nco` depends on `netcdf`

KateFriedman-NOAA replied:
fixed, thanks!

KateFriedman-NOAA commented:
@aerorahul Please recheck the modulefiles, I've reorganized them. Let me know if any are still in the wrong order, thanks!

@BoiVuong-NOAA I committed discussed updates to the util codes, please check I've made the updates correctly, thanks!

aerorahul reviewed (commented):
just one more place of module loads

KateFriedman-NOAA committed:
Moved nco module down after netcdf module load

- the nco module depends on netcdf, move it after netcdf module load

Refs: #399

BoiVuong-NOAA reviewed (approved):
All makefile files for util look good.

aerorahul commented on modulefiles/OznMonBuild.wcoss2:
```diff
@@ -0,0 +1,18 @@
+#%Module################################################
```
Is this used in the build?
I thought @EdwardSafford-NOAA mentioned that gfsv16.1 does not use this modulefile, but instead uses the one from GSI.

KateFriedman-NOAA replied:
Correct, it's not used per @EdwardSafford-NOAA's confirmation, so I removed the RadMon/OznMonBuild.wcoss2 I had initially added. Will use what comes from the GSI instead. I can retire the other platform copies of the RadMon/OznMonBuild files in the develop branch separately.

aerorahul reviewed (approved):
modulefiles look ok to me.

WalterKolczynski-NOAA commented on sorc/build_gfs_bufrsnd.sh:
```diff
@@ -21,6 +21,11 @@ if [ ! -d "../exec" ]; then
 fi
 
+if [ $target = "wcoss2" ]; then
+  export SIGIO_LIB4=$SIGIO_LIB
```
Aren't these usually set by the sigio module?

KateFriedman-NOAA replied:
Needed to support both stack (WCOSS2) and pre-stack (elsewhere).

WalterKolczynski-NOAA commented on sorc/build_tropcy_NEMS.sh:
```diff
@@ -29,6 +29,11 @@ else
 
 source ../modulefiles/modulefile.storm_reloc_v6.0.0.$target
+if [ $target = "wcoss2" ]; then
+  SIGIO_LIB4=$SIGIO_LIB
```
Ibid

KateFriedman-NOAA replied:
Needed to support both stack (WCOSS2) and pre-stack (elsewhere).

WalterKolczynski-NOAA commented on sorc/checkout.sh:
```diff
@@ -24,9 +24,10 @@ echo $topdir
 if [[ ! -d fv3gfs.fd ]] ; then
     rm -f ${topdir}/checkout-fv3gfs.log
-    git clone https://github.com/ufs-community/ufs-weather-model fv3gfs.fd >> ${topdir}/checkout-fv3gfs.log 2>&1
+    #git clone https://github.com/ufs-community/ufs-weather-model fv3gfs.fd >> ${topdir}/checkout-fv3gfs.log 2>&1
```
Remove commented code

KateFriedman-NOAA replied:
removed

WalterKolczynski-NOAA commented on util/sorc/mkgfsawps.fd/compile_mkgfsawps_wcoss.sh:
```diff
@@ -3,7 +3,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Is removing Hera as a target going to break anything there?

KateFriedman-NOAA replied:
Added Hera back in.

WalterKolczynski-NOAA commented on util/sorc/overgridid.fd/compile_overgridid_wcoss.sh:
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Ibid

KateFriedman-NOAA replied:
Added Hera back in.

WalterKolczynski-NOAA commented on util/sorc/rdbfmsua.fd/compile_rdbfmsua_wcoss.sh:
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Ibid

KateFriedman-NOAA replied:
Added Hera back in.

WalterKolczynski-NOAA commented on util/sorc/webtitle.fd/compile_webtitle_wcoss.sh:
```diff
@@ -10,7 +10,7 @@ LMOD_EXACT_MATCH=no
 cwd=`pwd`
 
-if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "hera" ] ; then
+if [ "$target" = "wcoss_dell_p3" ] || [ "$target" = "wcoss_cray" ] || [ "$target" = "wcoss2" ] ; then
```
Ibid

KateFriedman-NOAA replied:
Added Hera back in.

KateFriedman-NOAA committed:
Module updates in module_base.wcoss2

- add missing wgrib2 module load
- update grib_util module version to 1.2.3 and add libjpeg/9c module load for dependency
- add perl/5.32.0 module for Perl scripts, module was called pm5 on WCOSS-Dell

Refs: #399

BoiVuong-NOAA reviewed (approved):
looks good !

KateFriedman-NOAA committed:
Add Hera back into gfs_util compile scripts

Refs: #399

KateFriedman-NOAA committed:
Remove commented out lines in checkout.sh

Refs: #399

KateFriedman-NOAA committed:
Remove duplicate gldas entry in build config

Refs: #399

KateFriedman-NOAA committed:
Add WCOSS2 to gfs_util build check in build_all

Refs: #399

KateFriedman-NOAA commented:
Tested fresh clone and build_all of branch on Cactus, no issues, although I forgot to add WCOSS2 to the machine check for the gfs_util build in build_all.sh. Committed that @ ea5357a. Good to submit this.

KateFriedman-NOAA commented:
@BoiVuong-NOAA Please sync merge your fork copy of feature/ops-wcoss2 with the updated auth repo copy now. You should be able to use the "Fetch upstream" button. Let me know via email or Slack if you run into issues. Thanks!

BoiVuong-NOAA commented:
Thank You, Kate.
